File: backend/schemas/memory.py
"""学生学习记忆 API Schema。"""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field

from services.memory.schemas import MemoryEventInput, MemoryEventRecord


class MemoryEventCreateRequest(MemoryEventInput):
    """POST /api/memory/events 请求体。"""


class MemoryEventCreateResponse(BaseModel):
    ok: bool = True
    event: MemoryEventRecord


class LearningEvidenceItem(BaseModel):
    id: int
    event_type: str
    event_label: str = ""
    problem_slug: str = ""
    skill_id: str = ""
    chapter_id: str = ""
    summary: str = ""
    at: str | None = None


class MemorySummaryResponse(BaseModel):
    course_id: str
    learning_memory_summary: str = ""
    weak_patterns: list[str] = Field(default_factory=list)
    recent_count: int = 0
    dimension_evidence: dict[str, list[str]] = Field(default_factory=dict)
    update_reason: str = ""
    recent_evidence: list[LearningEvidenceItem] = Field(default_factory=list)
    generated_at: str = ""


class MemoryRecentResponse(BaseModel):
    items: list[MemoryEventRecord] = Field(default_factory=list)
    total: int = 0

File: frontend/src/api/memory.ts
import request from '@/utils/request'

import type { LearningEvidenceBrief } from '@/api/orchestrator'

export interface MemorySummary {
  course_id: string
  learning_memory_summary: string
  weak_patterns: string[]
  recent_count: number
  dimension_evidence: Record<string, string[]>
  update_reason: string
  recent_evidence: LearningEvidenceBrief[]
  generated_at: string
}

export interface GamePracticePayload {
  event_type: 'gamified_practice_complete'
  course_id: string
  chapter_id: string
  skill_id: string
  game_id: string
  level: string
  module_key: string
  success: boolean
  score: number
  attempts: number
  time_spent_seconds: number
  evidence_text: string
}

export async function fetchMemorySummary(params?: {
  course_id?: string
  limit?: number
}): Promise<MemorySummary> {
  return request.get('/api/memory/summary', {
    params: {
      course_id: params?.course_id ?? 'data_structures_algorithms',
      limit: params?.limit ?? 12,
    },
  }) as Promise<MemorySummary>
}

export async function recordGamePractice(payload: GamePracticePayload): Promise<void> {
  await request.post('/api/memory/events', {
    event_type: payload.event_type,
    course_id: payload.course_id,
    chapter_id: payload.chapter_id,
    skill_id: payload.skill_id,
    problem_slug: `game:${payload.game_id}:${payload.level}`,
    trace_summary: payload.evidence_text || `游戏 ${payload.game_id} 关卡 ${payload.level} ${payload.success ? '通关' : '未通关'}`,
    mastery_delta: payload.success ? 1 : 0,
    evidence_json: {
      game_id: payload.game_id,
      level: payload.level,
      module_key: payload.module_key,
      success: payload.success,
      score: payload.score,
      attempts: payload.attempts,
      time_spent_seconds: payload.time_spent_seconds,
      evidence_text: payload.evidence_text,
      persona_dimension: 'knowledge_base',
    },
  })
}
